python/ray/util/client/server/dataservicer.py

```diff
@@ -33,23 +48,47 @@ def Datapath(self, request_iterator, context):
             logger.error("Client connecting with no client_id")
             return
         logger.debug(f"New data connection from client {client_id}: ")
+        if not self._init(client_id, context):
+            return
         try:
-            for req in request_iterator:
+            request_queue = queue.Queue()
+            queue_filler_thread = Thread(
+                target=fill_queue,
+                daemon=True,
+                args=(request_iterator, request_queue))
+            queue_filler_thread.start()
+            """For non `async get` requests, this loop yields immediately
+            For `async get` requests, this loop:
+                 1) does not yield, it just continues
+                 2) When the result is ready, it yields
+            """
+            for req in iter(request_queue.get, None):
+                if isinstance(req, ray_client_pb2.DataResponse):
+                    # Early shortcut if this is the result of an async get.
+                    yield req
+                    continue
                 resp = None
                 req_type = req.WhichOneof("type")
                 if req_type == "init":
-                    resp = self._init(req.init, client_id)
-                    if resp is None:
-                        context.set_code(grpc.StatusCode.RESOURCE_EXHAUSTED)
-                        return
+                    resp_init = self.basic_service.Init(req.init)
+                    resp = ray_client_pb2.DataResponse(init=resp_init, )
                     logger.debug(f"Accepted data connection from {client_id}. "
                                  f"Total clients: {self.num_clients}")
                     accepted_connection = True
                 else:
                     assert accepted_connection
                     if req_type == "get":
-                        get_resp = self.basic_service._get_object(
-                            req.get, client_id)
+                        get_resp = None
+                        if req.get.asynchronous:
+                            get_resp = self.basic_service._async_get_object(
+                                req.get, client_id, req.req_id, request_queue)
+                            if get_resp is None:
+                                # The response for this request will be sent
+                                # when the object is ready.
+                                continue
+                        else:
+                            get_resp = self.basic_service._get_object(
+                                req.get, client_id)
                         resp = ray_client_pb2.DataResponse(get=get_resp)
                     elif req_type == "put":
                         put_resp = self.basic_service._put_object(
```

Comment:
Can we split these into two function calls one for asyncio and one for non asyncio?
I am trying to think of the best way to make this more readable. Maybe there are other ways to do it and I will refer here to your wisdom.

Reply:
@AmeerHajAli we cannot split `Datapath` into two separate functions. It is a stub method used by gRPC. I think leaving as is makes the most sense. Let me add an assert or two to make it clear what the logic is.

Reply:
I actually removed one level of nesting, because the `accepted_connection` boolean **should be set ** by the `self._init` function. The reason for this is that `accepted_connection` was used to ensure that we decrement `num_clients` if and only if we had incremented that. Now, if `num_clients` is **not incremented** (meaning we don't have resources), the function returns early.